File: docs/palindromes-in-peg/scaffold_places.py
"""Readonly Galil places: binary letters at odd places, virtual gaps at even.

There is no padded external input. A finite phase bit distinguishes a letter
from the gap immediately after it. InputHead still receives each real arrival
exactly once. A gap becomes available together with its preceding letter.
"""
from scaffold_input import InputHead


class PlaceHead:
  def __init__(self, vm, previous, builder, name):
    self.vm, self.builder, self.name = vm, builder, name
    self.head = InputHead(vm, previous, builder, name)
    self.gap = True if previous is None else vm.label(previous)[name + ".gap"]

  def append(self, cell):
    self.head.append(cell)

  def read(self):
    symbol = self.head.read()
    return None if symbol is None else "s" if self.gap else symbol

  def can_right(self):
    return not self.gap or self.head.can_right()

  def right(self):
    if self.gap:
      self.head.right()
    self.gap = not self.gap

  def left(self):
    if self.head.read() is None:
      raise ValueError("cannot move left of the input origin")
    if not self.gap:
      self.head.left()
    self.gap = not self.gap

  def is_first(self):
    return (not self.gap and not self.head.left_stack.empty()
            and self.head.left_stack.peek() is None)

  def copy_from(self, other):
    self.head.copy_from(other.head)
    self.gap = other.gap

  def finalize(self):
    self.head.finalize()
    self.builder.label[self.name + ".gap"] = self.gap

File: scala/pal/src/main/scala/pal/ScaffoldPlaces.scala
package pal

import Scavm.{Node, NodeRef, VM}
import ScavmStructs.Builder
import ScaffoldInput.{InputHead, ReadonlyHead}

/** Readonly Galil places: binary letters at odd places, virtual gaps at even.
  *
  * There is no padded external input. A finite phase bit distinguishes a letter
  * from the gap immediately after it. InputHead still receives each real arrival
  * exactly once. A gap becomes available together with its preceding letter.
  *
  * Python 原典: `scaffold_places.py`（`SCA_ONLINE_CONTROL.md` "Virtual places on
  * unchanged input"）。
  */
object ScaffoldPlaces {

  /** The symbol read at every gap place. */
  val GAP: String = "s"

  /** One phase bit `gap` over an [[InputHead]]: place `2i+1` is letter `i` with
    * `gap = false`, place `2i+2` is the gap after it with `gap = true`; the origin
    * (place 0) is the head's origin with `gap = true`.
    */
  final class PlaceHead(val vm: VM, previous: Option[Node], val builder: Builder, val name: String)
      extends ReadonlyHead[PlaceHead] {

    val head: InputHead = new InputHead(vm, previous, builder, name)

    var gap: Boolean = previous.fold(true)(p => vm.label(p)(name + ".gap").asBool)

    def append(cell: NodeRef): Unit = head.append(cell)

    /** The letter, [[GAP]] at a gap place, `None` at the origin. */
    def read(): Option[String] = head.read().map(symbol => if (gap) { GAP } else { symbol })

    def canRight: Boolean = !gap || head.canRight

    def right(): Unit = {
      if (gap) {
        head.right()
      }
      gap = !gap
    }

    def left(): Unit = {
      if (head.read().isEmpty) {
        throw new IllegalArgumentException("cannot move left of the input origin")
      }
      if (!gap) {
        head.left()
      }
      gap = !gap
    }

    /** Whether this head is at place 1, the first letter. */
    def isFirst: Boolean = !gap && !head.leftStack.empty && head.leftStack.peek.isEmpty

    def copyFrom(other: PlaceHead): Unit = {
      head.copyFrom(other.head)
      gap = other.gap
    }

    def finish(): Unit = {
      head.finish()
      builder.label(name + ".gap") = gap
    }
  }
}
